File: servicios/servicio_auditoria/app/Domain/Audit/AuditEvent.php
<?php

namespace App\Domain\Audit;

use InvalidArgumentException;


final class AuditEvent
{
    public function __construct(

        public readonly string $eventId,

        public readonly int $eventVersion,

        public readonly string $occurredAt,

        public readonly string $service,

        public readonly string $module,

        public readonly string $action,

        public readonly string $result,

        public readonly ?string $actorUserId,

        public readonly ?string $actorName,

        public readonly ?string $actorRole,

        public readonly ?string $entityType,

        public readonly ?string $entityId,

        public readonly ?string $correlationId,

        public readonly ?string $ipAddress,

        public readonly ?string $userAgent,

        public readonly ?string $description,

        public readonly ?string $reason,

        public readonly array $detail,

        public readonly array $changes,

    ) {
    }


    // ======================================================
    // CREAR DESDE PAYLOAD
    // ======================================================

    public static function fromArray(
        array $payload
    ): self {

        $eventId = trim(
            (string) (
                $payload['event_id']
                ?? ''
            )
        );


        if (
            !self::isUuid(
                $eventId
            )
        ) {

            throw new InvalidArgumentException(
                'event_id debe ser un UUID válido.'
            );

        }


        $eventVersion = (int) (
            $payload['event_version']
            ?? 0
        );


        if (
            $eventVersion !== 1
        ) {

            throw new InvalidArgumentException(
                'Versión de evento no soportada.'
            );

        }


        $service = self::requiredCode(
            $payload['service']
            ?? null,
            'service'
        );


        $module = self::requiredCode(
            $payload['module']
            ?? null,
            'module'
        );


        $action = self::requiredCode(
            $payload['action']
            ?? null,
            'action'
        );


        $result = self::requiredCode(
            $payload['result']
            ?? null,
            'result'
        );


        $actor = (
            is_array(
                $payload['actor']
                ?? null
            )
                ? $payload['actor']
                : []
        );


        $entity = (
            is_array(
                $payload['entity']
                ?? null
            )
                ? $payload['entity']
                : []
        );


        $context = (
            is_array(
                $payload['context']
                ?? null
            )
                ? $payload['context']
                : []
        );


        return new self(

            eventId:
                $eventId,

            eventVersion:
                $eventVersion,

            occurredAt:
                trim(
                    (string) (
                        $payload['occurred_at']
                        ?? gmdate('c')
                    )
                ),

            service:
                $service,

            module:
                $module,

            action:
                $action,

            result:
                $result,

            actorUserId:
                self::optionalText(
                    $actor['user_id']
                    ?? null
                ),

            actorName:
                self::optionalText(
                    $actor['name']
                    ?? null
                ),

            actorRole:
                self::optionalText(
                    $actor['role']
                    ?? null
                ),

            entityType:
                self::optionalUpper(
                    $entity['type']
                    ?? null
                ),

            entityId:
                self::optionalText(
                    $entity['id']
                    ?? null
                ),

            correlationId:
                self::optionalText(
                    $payload['correlation_id']
                    ?? null
                ),

            ipAddress:
                self::optionalText(
                    $context['ip_address']
                    ?? null
                ),

            userAgent:
                self::optionalText(
                    $context['user_agent']
                    ?? null
                ),

            description:
                self::optionalText(
                    $payload['description']
                    ?? null
                ),

            reason:
                self::optionalText(
                    $payload['reason']
                    ?? null
                ),

            detail:
                (
                    is_array(
                        $payload['detail']
                        ?? null
                    )
                        ? $payload['detail']
                        : []
                ),

            changes:
                (
                    is_array(
                        $payload['changes']
                        ?? null
                    )
                        ? $payload['changes']
                        : []
                ),

        );
    }


    // ======================================================
    // CÓDIGO OBLIGATORIO
    // ======================================================

    private static function requiredCode(
        mixed $value,
        string $field
    ): string {

        $value = strtoupper(
            trim(
                (string) $value
            )
        );


        if (
            $value === ''
        ) {

            throw new InvalidArgumentException(
                "El campo {$field} es obligatorio."
            );

        }


        return $value;
    }


    // ======================================================
    // TEXTO OPCIONAL
    // ======================================================

    private static function optionalText(
        mixed $value
    ): ?string {

        if (
            $value === null
        ) {
            return null;
        }


        $value = trim(
            (string) $value
        );


        return (
            $value !== ''
                ? $value
                : null
        );
    }


    // ======================================================
    // CÓDIGO OPCIONAL
    // ======================================================

    private static function optionalUpper(
        mixed $value
    ): ?string {

        $value = self::optionalText(
            $value
        );


        if (
            $value === null
        ) {

            return null;

        }


        return strtoupper(
            $value
        );
    }


    // ======================================================
    // UUID
    // ======================================================

    private static function isUuid(
        string $value
    ): bool {

        return (

            preg_match(

                '/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$/',

                $value

            )

            === 1

        );
    }
}

File: servicios/servicio_usuarios/identidad/domain/events/audit_event.py
import uuid

from dataclasses import (
    asdict,
    dataclass,
    field,
)

from datetime import (
    datetime,
    timezone,
)

from typing import (
    Any,
)


# ==========================================================
# EVENTO DE AUDITORÍA
# ==========================================================

@dataclass(
    frozen=True
)
class AuditEvent:
    """
    Contrato estándar de eventos de auditoría.

    Este objeto pertenece al dominio del microservicio
    productor.

    NO conoce:
    - PostgreSQL Auditoría
    - Laravel
    - HTTP
    - tablas externas

    Solamente describe un hecho ocurrido.
    """

    event_id: str

    event_version: int

    occurred_at: str

    service: str

    module: str

    action: str

    result: str

    actor: dict[
        str,
        Any,
    ] = field(
        default_factory=dict
    )

    entity: dict[
        str,
        Any,
    ] = field(
        default_factory=dict
    )

    correlation_id: (
        str
        | None
    ) = None

    context: dict[
        str,
        Any,
    ] = field(
        default_factory=dict
    )

    description: (
        str
        | None
    ) = None

    reason: (
        str
        | None
    ) = None

    detail: dict[
        str,
        Any,
    ] = field(
        default_factory=dict
    )

    changes: list[
        dict[
            str,
            Any,
        ]
    ] = field(
        default_factory=list
    )


    # ======================================================
    # CREAR EVENTO
    # ======================================================

    @classmethod
    def create(
        cls,
        *,
        service,
        module,
        action,
        result,

        actor_user_id=None,
        actor_name=None,
        actor_role=None,

        entity_type=None,
        entity_id=None,

        correlation_id=None,

        ip_address=None,
        user_agent=None,

        description=None,
        reason=None,

        detail=None,
        changes=None,
    ):

        return cls(

            event_id=
                str(
                    uuid.uuid4()
                ),

            event_version=
                1,

            occurred_at=
                datetime.now(
                    timezone.utc
                )
                .isoformat(),

            service=
                cls._required_upper(
                    service,
                    "service",
                ),

            module=
                cls._required_upper(
                    module,
                    "module",
                ),

            action=
                cls._required_upper(
                    action,
                    "action",
                ),

            result=
                cls._required_upper(
                    result,
                    "result",
                ),

            actor={
                "user_id":
                    cls._optional_text(
                        actor_user_id
                    ),

                "name":
                    cls._optional_text(
                        actor_name
                    ),

                "role":
                    cls._optional_text(
                        actor_role
                    ),
            },

            entity={
                "type":
                    cls._optional_upper(
                        entity_type
                    ),

                "id":
                    cls._optional_text(
                        entity_id
                    ),
            },

            correlation_id=
                cls._optional_text(
                    correlation_id
                ),

            context={
                "ip_address":
                    cls._optional_text(
                        ip_address
                    ),

                "user_agent":
                    cls._optional_text(
                        user_agent
                    ),
            },

            description=
                cls._optional_text(
                    description
                ),

            reason=
                cls._optional_text(
                    reason
                ),

            detail=(
                dict(
                    detail
                )
                if isinstance(
                    detail,
                    dict,
                )
                else {}
            ),

            changes=(
                list(
                    changes
                )
                if isinstance(
                    changes,
                    list,
                )
                else []
            ),

        )


    # ======================================================
    # ROUTING KEY
    # ======================================================

    @property
    def routing_key(
        self,
    ):

        return (
            f"{self.service.lower()}"
            ".audit.v1"
        )


    # ======================================================
    # SERIALIZACIÓN
    # ======================================================

    def to_dict(
        self,
    ):

        return asdict(
            self
        )


    # ======================================================
    # UTILIDADES
    # ======================================================

    @staticmethod
    def _required_upper(
        value,
        field_name,
    ):

        value = str(
            value or ""
        ).strip()

        if not value:

            raise ValueError(
                (
                    "El campo "
                    f"{field_name} "
                    "es obligatorio."
                )
            )

        return value.upper()


    @staticmethod
    def _optional_upper(
        value,
    ):

        if value is None:

            return None

        value = str(
            value
        ).strip()

        return (
            value.upper()
            if value
            else None
        )


    @staticmethod
    def _optional_text(
        value,
    ):

        if value is None:

            return None

        value = str(
            value
        ).strip()

        return (
            value
            if value
            else None
        )